CUDA/GB_cuda_stream_pool.cu

```diff
@@ -0,0 +1,107 @@
+#include "GB_cuda_stream_pool.hpp"
+
+#define STREAMS_PER_DEVICE 32
+
+struct GB_cuda_stream_pool
+{
+    std::vector<std::array<cudaStream_t, STREAMS_PER_DEVICE>> streams ;
+    std::vector<int> nstreams_avail;
+} ;
+
+static GB_cuda_stream_pool pool ;
+
+#undef GB_FREE_ALL
+#define GB_FREE_ALL                                     \
+{                                                       \
+    while (pool.streams.size())                         \
+    {                                                   \
+        auto curr = pool.streams.back() ;               \
+        int end = pool.nstreams_avail.back() - 1 ;      \
+        for (int idx = end; idx >= 0 ; idx--)           \
+        {                                               \
+            cudaStreamDestroy (curr[idx]) ;             \
+        }                                               \
+        pool.streams.pop_back() ;                       \
+        pool.nstreams_avail.pop_back() ;                \
+    }                                                   \
+}
+
+void GB_cuda_release_stream (int device, cudaStream_t *stream)
+{
+    if (stream == nullptr || (*stream) == nullptr)
+    {
+        return ;
+    }
+
+    ASSERT (device < pool.streams.size()) ;
+
+    #pragma omp critical
+    {
+        if (pool.nstreams_avail[device] == STREAMS_PER_DEVICE)
+        {
+            // Pool is full; destroy the stream
+            cudaStreamDestroy (*stream) ;
+        }
+        else
+        {
+            // Check the stream back in; it's OK if the stream wasn't
+            // created at init time. Whatever stream was will be destroyed
+            // when it is checked back in.
+            size_t stream_idx = pool.nstreams_avail[device];
+            pool.streams[device][stream_idx] = (*stream) ;
+            pool.nstreams_avail[device]++ ;
+        }
+    }
+    (*stream) = nullptr ;
+}
+
+GrB_Info GB_cuda_grab_stream (int device, cudaStream_t *stream)
+{
+    ASSERT (stream != nullptr) ;
+    ASSERT (device < pool.streams.size()) ;
+    GrB_Info ret = GrB_SUCCESS ;
+
+    #pragma omp critical
+    {
+        if (!pool.nstreams_avail[device])
+        {
+            // Pool is empty; create a stream
+            GB_cuda_set_device (device) ;
```

Comment:
We should set the device back to the original one, after creating the stream.

Reply:
My intention with the grab/release was that we should always pass in the current device, so we wouldn't need to set the device back.
Looking at this again, it's probably better to remove the device as an argument and just get the current device inside grab/release. This is the only use case we need and aligns with how the CUDA APIs behave. Does that sound good?

Reply:
yes, that sounds like a better strategy to me.

Reply:
Sounds good, I'm out of town currently and don't have access to my normal setup. If there are no other issues, can you merge this in and make the change separately?